DPE: Add support for ReadOnly attributes

amzn-tmryan opened the pull request:
**Description**
This PR adds support to the DPE for the `AZ::Edit::Attributes::Readonly` attribute. Nodes with this attribute set will be disabled when their property handler widget is created in `DPERowWidget::CreateWidgetForHandler`.

Any existing ReadOnly attribute will be detected per node and its value stored in the node's disabled flag prior to caching attributes in the LegacyReflectionBridge. When caching attributes, we already check if the node should be disabled (as is the case with set data structures) hence we roll the legacy ReadOnly attribute into our DPE Disabled attribute. Also, child nodes of disabled parents will inherit the Disabled attribute and elements of ReadOnly containers will disable their container action buttons.

Box 1 in the image below shows that non-readonly set elements are unaffected by this change. Box 2 shows that the legacy ReadOnly attribute can be applied to simple property editor nodes and to entire containers:


![dpe_disabledEditors](https://user-images.githubusercontent.com/104796591/194473713-373c7036-3396-4510-a0be-a9c551cc0f4c.jpg)

**Updates since originally approved**
- Added support for applying/removing the Disabled and AncestorDisabled attributes at runtime
  - There is now a `CallbackAttributeDefinition` called `SetDisabled` that is added to property editors built in the `ReflectionAdapter`
  - There is also a handler for that callback attribute in `ReflectionAdapter::HandleMessage`, in the lambda `handlePropertyEditorDisabledChanged`, that handles the three main cases we run into:
    - disabling a node and propagating to its descendants
    - enabling a node and removing the ancestor disabled state of its descendants
    - disabling an element of a container
- Refactored the attribute updating code inside of `DPERowWidget::SetPropertyEditorAttributes` in order to fix some troubles with refreshing re-enabled nodes in the view

Here you can see the state changes propagating in the view and their respective DOM changes (expand to view properly):
![dpeReadOnlyRuntimeInheritance](https://user-images.githubusercontent.com/104796591/199871132-af1a307f-9ebf-42d9-9934-2fb433717b4c.gif)

**Testing**
- Added some ReadOnly nodes to the standalone debug DPE app and verified that they are non-editable
- Verified that non-ReadOnly set containers functionality is unchanged by this PR
- Created a temp button that toggled the disabled state of a node and verified that the two disabled states are propagated correctly and interact with each other correctly (a disabled node should remain so when its parent becomes enabled, etc.)
- Verified there were no unexpected errors/assertions in a debug build of the standalone debug DPE

Signed-off-by: Ryan <ryanto@amazon.com>

alexmontAmazon reviewed (commented):
My main question here is whether we even need a readOnly attribute if there is already a disabled attribute that would do the same thing. Is "ReadOnly" already an attribute that can come through serialize context or the like? Also, @dtamkin1, looks like we need to add Description and Disabled to the list of attributes we need to handle in patches...

amzn-tmryan commented:
> My main question here is whether we even need a readOnly attribute if there is already a disabled attribute that would do the same thing. Is "ReadOnly" already an attribute that can come through serialize context or the like? Also, @dtamkin1, looks like we need to add Description and Disabled to the list of attributes we need to handle in patches...

Chatted with Daniel. I'm not sure if ReadOnly is a legacy attribute or not, but maybe we can combine them in the legacy bridge so the DPE only needs to understand its Disabled attribute.

Will get another commit up for that.

amzn-tmryan commented:
Updated commit to convert an existing legacy ReadOnly attribute into a DPE Disabled attribute.

amzn-tmryan commented:
Simplified the code a bit and removed some unnecessary stuff.

amzn-tmryan commented:
Actually, this doesn't work for containers that are declared as ReadOnly. The children are still editable. Investigating.

alexmontAmazon commented:
> Actually, this doesn't work for containers that are declared as ReadOnly. The children are still editable. Investigating.

@amzn-tmryan, I think it's reasonable to recurse into child rows, either where we handle it in DocumentPropertyEditor, or when it's built in LegacyReflectionBridge

amzn-tmryan commented:
Added a commit to disable container elements. Technically, it makes all child nodes inherit the "disabledness" of their parent which is likely a container, but I supposed it doesn't have to be.

The problem I have with this solution might be a naming issue? I see how the behavior of Disabled and ReadOnly attributes is the same in the DPE, and that the big difference between disabling set elements versus a read only container's elements depends on whether their parent is disabled or not, but it really feels like I'm forcing everything to fit together here.

Critical feedback appreciated!

amzn-tmryan commented on Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/Reflection/LegacyReflectionBridge.cpp:
```diff
@@ -159,6 +159,7 @@ namespace AZ::Reflection
                 bool m_skipLabel = false;
                 AZStd::string m_labelOverride;
                 bool m_disableEditor = false;
+                bool m_disableContainer = false;
```
This name doesn't really make sense. It's not really specific to containers, but to nodes with the ReadOnly attribute. Was trying to avoid using ReadOnly since we could just register the ReadOnly attribute in that case.

amzn-tmryan replied:
Talked with Alex again and made some sense of things. At first I was using the same attribute to determine whether containers and elements should have their action buttons disabled, but that was incorrect and causing confusion on my part.

If there is a better way to do this @NicholasVanSickle then please let me know.

lumberyard-employee-dm commented:
Should we get UX involved here?

amzn-tmryan commented:
> Should we get UX involved here?

We could, but we'll need to work out what's most important for them to weigh-in on first. The troubles with this PR have been coding/design related so far.

amzn-tmryan commented:
Talked with @alexmontAmazon and we shouldn't need any UX perspective on this particular change since we are copying the UX from the RPE which was already UX approved.

lumberyard-employee-dm commented on Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/Reflection/LegacyReflectionBridge.cpp:
```diff
@@ -263,6 +264,22 @@ namespace AZ::Reflection
+                // If our parent node is disabled then we should inherit its Disabled attribute and make note of the inheritance
+                if (parentData.m_disableEditor)
+                {
+                    nodeData->m_disableEditor = true;
+                    nodeData->m_isAncestorDisabled = true;
```
This seems like the m_isAncestorDisabled would be stale if the parent node is then enabled.

lumberyard-employee-dm commented on Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/PropertyEditorNodes.h:
```diff
@@ -106,6 +106,10 @@ namespace AZ::DocumentPropertyEditor::Nodes
         static constexpr auto ValueHashed = AttributeDefinition<AZ::Uuid>("ValueHashed");
 
+        //! In some cases, a node may need to know that it is descended from a disabled ancestor. For example, disabled
+        //! elements of a disabled container might requrie different treatment than disabled elements of an enabled container.
```
typo: `requrie`
```suggestion
        //! elements of a disabled container might require different treatment than disabled elements of an enabled container.
```

amzn-tmryan commented:
Added support for propagating disabled state changes at runtime. Updated the description with details and added a gif of the debug standalone DPE in action (temp disable buttons have been removed for this PR).

cgalvan commented on Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/PropertyEditorNodes.h:
```diff
@@ -103,9 +103,15 @@ namespace AZ::DocumentPropertyEditor::Nodes
+        static constexpr auto Disabled = AttributeDefinition<bool>("Disabled");
+        //! In some cases, a node may need to know that it is descended from a disabled ancestor. For example, disabled
+        //! elements of a disabled container might require different treatment than disabled elements of an enabled container.
+        static constexpr auto AncestorDisabled = AttributeDefinition<bool>("AncestorDisabled");
```
Should this be private? This looks like an internal state rather than an attribute the user would be setting, correct?

amzn-tmryan replied:
It's true that we use this attribute internally, but I'm not sure if we want to preclude users from using it if they found a need.

Also, I'm not sure the DPE design has room for private attributes as is. @alexmontAmazon might be able to shed some more light on that or correct me if I'm wrong.

alexmontAmazon replied:
the way Nick designed this, the PropertyEditorNodes is only a container for attributes, so they don't really consume them themselves, therefore this must be public. It makes sense to keep the handling of this kind of attribute private to the LegacyReflectionBridge, which is already the case. On a related note, the DPE class itself has waaaay too many public methods just so the DPERowWidget can access them, which is poor design. I'm going to make the row a friend of the DPE and then to do a pass making methods private/protected.

alexmontAmazon commented on Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:
```diff
@@ -610,164 +610,76 @@ namespace AzToolsFramework
+        if (attributeIter != m_domOrderToAttributeInfo.end())
         {
-            priorColumnIndex = m_columnLayout->indexOf(m_domOrderedChildren[searchIndex]);
+            currentInfo = m_domOrderToAttributeInfo[domIndex];
```
not your original code I know, but could you change this to "m_childToAttributeInfo" please?

amzn-tmryan replied:
Done.

alexmontAmazon reviewed (commented):
as we discussed, let's move the disabled setting methods to the adapter level so that they're not on each node, which will also allow row-level disabling. Otherwise, all looks good!

amzn-tmryan commented:
Addressed comments, moved the disabled attribute types to the `NodeWithVisiblityControl` level so they are available to more node types, and changed the handler for updating a node's disabled state to an adapter level handler that takes a target node path (for less DOM clutter).

amzn-tmryan commented on Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp:
```diff
@@ -624,6 +643,177 @@ namespace AZ::DocumentPropertyEditor
+
+                    // We can stop traversing this path if the node has a truthy disabled attribute since its descendants
+                    // should retain their inherited disabled state
+                    if (auto iter = childNode->FindMember(disabledAttributeName); iter != childNode->MemberEnd() && iter->second.GetBool())
```
Done!

amzn-tmryan commented on Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp:
```diff
@@ -347,6 +349,12 @@ namespace AZ::DocumentPropertyEditor
                     m_builder.Attribute(Nodes::PropertyEditor::UseMinimumWidth, true);
                     m_builder.Attribute(Nodes::PropertyEditor::Alignment, Nodes::PropertyEditor::Align::AlignRight);
                     m_builder.Attribute(Nodes::ContainerActionButton::Action, Nodes::ContainerAction::RemoveElement);
+                    auto disabledValue = attributes.Find(DocumentPropertyEditor::Nodes::PropertyEditor::AncestorDisabled.GetName());
```
Done.

amzn-tmryan commented on Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp:
```diff
@@ -643,140 +643,88 @@ namespace AZ::DocumentPropertyEditor
 
-            // Applies the attribute change to any descendants in unvisitedChildren until its done
-            const auto propagateAttributeChangeToDescendants = [&](AZStd::function<void(Dom::Path&)> procedure)
+            // This lambda applies the attribute change to any descendants in unvisitedChildren until its done
```
Typo here: unvisitedChildren -> unvisitedDescendants
Should probably just call it something less variable name-y; will pick that up in another PR if AR succeeds and there aren't any other comment.s
